pyglet/experimental/multitexture_sprite.py

```diff
@@ -0,0 +1,331 @@
+
+    # Generate the default vertex shader
+    vertex_source = []
+    vertex_source.append("#version 150 core")
+    vertex_source.append("in vec3 translate;")
+    vertex_source.append("in vec4 colors;")
+    vertex_source.append("in vec2 scale;")
+    vertex_source.append("in vec3 position;")
+    vertex_source.append("in float rotation;")
+    vertex_source.extend([f"in vec3 {name}_coords;" for name in images.keys()])
+
+    vertex_source.append("out vec4 vertex_colors;")
+    vertex_source.extend([f"out vec3 {name}_coords_frag;" for name in images.keys()])
+
+    vertex_source.append("uniform WindowBlock {mat4 projection; mat4 view;} window;")
+    vertex_source.append("mat4 m_scale = mat4(1.0);")
+    vertex_source.append("mat4 m_rotation = mat4(1.0);")
+    vertex_source.append("mat4 m_translate = mat4(1.0);")
+
+    vertex_source.append("void main() {")
+    vertex_source.append("  m_scale[0][0] = scale.x;")
+    vertex_source.append("  m_scale[1][1] = scale.y;")
+    vertex_source.append("  m_translate[3][0] = translate.x;")
+    vertex_source.append("  m_translate[3][1] = translate.y;")
+    vertex_source.append("  m_translate[3][2] = translate.z;")
+    vertex_source.append("  m_rotation[0][0] =  cos(-radians(rotation));")
+    vertex_source.append("  m_rotation[0][1] =  sin(-radians(rotation));")
+    vertex_source.append("  m_rotation[1][0] = -sin(-radians(rotation));")
+    vertex_source.append("  m_rotation[1][1] =  cos(-radians(rotation));")
+    vertex_source.append("  gl_Position = window.projection * window.view * m_translate * m_rotation * m_scale * vec4(position, 1.0);")
+    vertex_source.append("  vertex_colors = colors;")
+
+    vertex_source.extend([f"{name}_coords_frag = {name}_coords;" for name in images.keys()])
+    vertex_source.append("}")
+    vertex_source = '\n'.join(vertex_source)
+
+    fragment_source = []
+    fragment_source.append("#version 150 core")
+    fragment_source.append("in vec4 vertex_colors;")
+    fragment_source.extend([f"in vec3 {name}_coords_frag;" for name in images.keys()])
+
+    fragment_source.append("out vec4 final_colors;")
+
+    fragment_source.extend([f"uniform {_SAMPLER_TYPES[tex.target]} {name};" for name,tex in images.items()])
+
+    fragment_source.append("vec4 layer(vec4 foreground, vec4 background) {")
+    fragment_source.append("  return foreground * foreground.a + background * (1.0 - foreground.a);")
+    fragment_source.append("}")
+
+    fragment_source.append("void main() {")
+    fragment_source.append("  vec4 color = vec4(0.0, 0.0, 0.0, 1.0);")
+    fragment_source.extend([f"  color = layer(texture({name}, {name}_coords_frag{_SAMPLER_COORDS[tex.target]}), color);" for name,tex in images.items()])
+    fragment_source.append("  final_colors = color * vertex_colors;")
+    fragment_source.append("}")
+
+    fragment_source = '\n'.join(fragment_source)
```

Comment:
This area looks a lot cleaner.

Perhaps just inject what needs to be changed, so you can keep it all one string?

For example:
```py
in_tex_coords = "\n".join([f"in vec3 {name}_coords;" for name in images])

frag = f"""
#version 150 core
in vec3 translate;
in vec4 colors;
in vec2 scale;
in vec3 position;
in float rotation;
{in_tex_coords}
 """

print(frag)
```

Reply:
Good idea.  Implemented requested change.